File: src/features/typing/data/java/top_k_frequent_elements.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> count = new HashMap<>();
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }
        List<List<Integer>> buckets = new ArrayList<>();
        for (int i = 0; i <= nums.length; i++) {
            buckets.add(new ArrayList<>());
        }
        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            buckets.get(entry.getValue()).add(entry.getKey());
        }
        List<Integer> picked = new ArrayList<>();
        for (int i = buckets.size() - 1; i >= 0 && picked.size() < k; i--) {
            picked.addAll(buckets.get(i));
        }
        int[] result = new int[picked.size()];
        for (int i = 0; i < picked.size(); i++) {
            result[i] = picked.get(i);
        }
        return result;
    }
}

File: src/features/typing/data/dart/top_k_frequent_elements.dart
class Solution {
  List<int> topKFrequent(List<int> nums, int k) {
    final count = <int, int>{};
    for (final num in nums) {
      count[num] = (count[num] ?? 0) + 1;
    }
    final buckets = List<List<int>>.generate(nums.length + 1, (_) => <int>[]);
    for (final entry in count.entries) {
      buckets[entry.value].add(entry.key);
    }
    final result = <int>[];
    for (int i = buckets.length - 1; i >= 0 && result.length < k; i--) {
      result.addAll(buckets[i]);
    }
    return result;
  }
}
